actionpack/lib/action_dispatch/routing/mapper.rb

```diff
@@ -1168,13 +1168,27 @@ module Resources
         CANONICAL_ACTIONS = %w(index create new show update destroy)
 
         class Resource # :nodoc:
+          class << self
+            def default_actions(api_only)
+              if api_only
+                [:index, :create, :show, :update, :destroy]
+              else
+                [:index, :create, :new, :show, :update, :destroy, :edit]
+              end
+            end
+          end
+
           attr_reader :controller, :path, :param
 
           def initialize(entities, api_only, shallow, options = {})
             if options[:param].to_s.include?(":")
               raise ArgumentError, ":param option can't contain colons"
             end
 
+            if (invalid_actions = (options.values_at(:only, :except).flatten.compact - default_actions)).any?
+              raise ArgumentError, "expected :only and :except to be one or more of #{default_actions}, got #{invalid_actions}"
```

Comment:
`(options.values_at(:only, :except).flatten.compact - default_actions)` could be extracted to a memoized `invalid_actions` method. After that we can use it on both places.

Reply:
This is only used once when creating the resource, what's the benefit in memoizing it?

> After that we can use it on both places.

Not sure what you mean by this. Both a `Resource` and `SingletonResource` execute this just once on initialisation (see https://github.com/rails/rails/blob/cacf96ecd053f0a3d64c7bf478dac63f98e6aa6d/actionpack/lib/action_dispatch/routing/mapper.rb#L1272).

Reply:
I see. If you want, you can still encapsulate the code without the memoization. You can document the code with a good method name and separate responsibilities.

Reply:
Fair call, I've addressed in [these changes](https://github.com/rails/rails/compare/1b40f99aebbec5a520c50f56b9ac5a8338b1153a..6692c5ec7c7fee8b18884b3c542137dd3231a8c3).

In doing so I realized that the validation logic didn't account for `api_only` when comparing against the default actions, so I fixed that. This resulted in some test failures in action mailbox since it generates some resource routes and doesn't account for whether the app is api only or not when using `only:`, I've fixed that here too.